error/s2n_errno.c

```diff
@@ -31,209 +31,238 @@ static const char *no_such_language = "Language is not supported for error trans
+    ADD_ERROR(S2N_ERR_TICKET_KEY_NOT_UNIQUE, "Cannot add session ticket key because it was added before"),
+    ADD_ERROR(S2N_ERR_TICKET_KEY_LIMIT, "Limit reached for unexpired session ticket keys"),
+    ADD_ERROR(S2N_ERR_NO_TICKET_ENCRYPT_DECRYPT_KEY, "No key in encrypt-decrypt state is available to encrypt session ticket"),
+    ADD_ERROR(S2N_ERR_ENCRYPT_DECRYPT_KEY_SELECTION_FAILED, "Failed to select a key from keys in encrypt-decrypt state"),
+    ADD_ERROR(S2N_ERR_ENCRYPT_DECRYPT_KEY_SELECTION_FAILED, "Failed to select a key from keys in encrypt-decrypt state"),
+    ADD_ERROR(S2N_ERR_KEY_USED_IN_SESSION_TICKET_NOT_FOUND, "Key used in already assigned session ticket not found for decryption"),
+    ADD_ERROR(S2N_ERR_SENDING_NST, "Error in session ticket status encountered before sending NST"),
+    ADD_ERROR(S2N_ERR_INVALID_DYNAMIC_THRESHOLD, "invalid dynamic record threshold"),
+    ADD_ERROR(S2N_ERR_INVALID_ARGUMENT, "invalid argument provided into a function call"),
+    ADD_ERROR(S2N_ERR_NOT_IN_UNIT_TEST, "Illegal configuration, can only be used during unit tests"),
+    ADD_ERROR(S2N_ERR_UNSUPPORTED_CPU, "Unsupported CPU architecture"),
+    ADD_ERROR(S2N_ERR_SESSION_ID_TOO_SHORT, "Session id is too short"),
+    ADD_ERROR(S2N_ERR_CONNECTION_CACHING_DISALLOWED, "This connection is not allowed to be cached"),
+    ADD_ERROR(S2N_ERR_SESSION_TICKET_NOT_SUPPORTED, "Session ticket not supported for this connection"),
+    ADD_ERROR(S2N_ERR_OCSP_NOT_SUPPORTED, "OCSP stapling was requested, but is not supported"),
 };
-const int num_of_errors = sizeof(S2N_ERROR_EN) / sizeof(S2N_ERROR_EN[0]);
-static struct s2n_map *error_translation_table = NULL;
 
-typedef union {
-    uint8_t bytes[sizeof(int)];
-    int raw;
-} key_type;
+uint32_t S2N_ERROR_COUNT[S2N_ERR_TYPE_COUNT] = {
+    [S2N_ERR_T_OK] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_OK),
+    [S2N_ERR_T_IO] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_IO),
+    [S2N_ERR_T_CLOSED] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_CLOSED),
+    [S2N_ERR_T_BLOCKED] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_BLOCKED),
+    [S2N_ERR_T_ALERT] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_ALERT),
+    [S2N_ERR_T_PROTO] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_PROTO),
+    [S2N_ERR_T_INTERNAL] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_INTERNAL),
+    [S2N_ERR_T_USAGE] = S2N_ERROR_TYPE_ERROR_COUNT(S2N_ERR_T_USAGE),
+};
 
-typedef union {
-    uint8_t bytes[sizeof(uintptr_t)];
-    uintptr_t raw;
-} value_type;
+struct s2n_error_translation *S2N_ERROR_EN[S2N_ERR_TYPE_COUNT] = {
+    [S2N_ERR_T_OK] = S2N_ERR_T_OK_EN,
+    [S2N_ERR_T_IO] = S2N_ERR_T_IO_EN,
+    [S2N_ERR_T_CLOSED] = S2N_ERR_T_CLOSED_EN,
+    [S2N_ERR_T_BLOCKED] = S2N_ERR_T_BLOCKED_EN,
+    [S2N_ERR_T_ALERT] = S2N_ERR_T_ALERT_EN,
+    [S2N_ERR_T_PROTO] = S2N_ERR_T_PROTO_EN,
+    [S2N_ERR_T_INTERNAL] = S2N_ERR_T_INTERNAL_EN,
+    [S2N_ERR_T_USAGE] = S2N_ERR_T_USAGE_EN,
+};
```

Comment:
This is creative.

Reply:
Why not make it a switch/case? That should also be O(1) assuming the compiler does the right thing.

Reply:
The advantage of this approach is that if you're missing a case in the switch, you get a warning.

```
enum error_cases {
      A,
      B,
      C,
      D
};

#define ERROR_CASE( err, string) case err: do { return string;} while (0)

const char* error_string(enum error_cases error)
{
  switch(error) {
    ERROR_CASE(A, "a");
    ERROR_CASE(B, "b");
    ERROR_CASE(C, "c");
    ERROR_CASE(D, "d");
  }
  return "default case";
  
}

int main() {
  printf("%s\n", error_string(A));
  printf("%s\n", error_string(B));
  printf("%s\n", error_string(C));
  printf("%s\n", error_string(D));
}
```

Reply:
Just for fun, I took a look at the compiler output https://godbolt.org/z/8Ry42y  Given reasonable optimization options, it does the right O(1) thing.

Reply:
Yup, makes sense.

Reply:
Modified your example a bit to include gaps in enum, like we do with s2n_errors. [gcc](https://godbolt.org/z/FKeXtA) does cmps to figure out which interval are we in and then proceeds to index. [clang](https://godbolt.org/z/V7EsbB) does something similar, but it creates a ton of jumps instead of indexing.

Either way I think for our purposes switch statement should be fast enough given optimization and is better at catching errors, so I'll make a change.